File: src/main/java/juuxel/woodsandmires/feature/ShrubFeatureConfig.java
package juuxel.woodsandmires.feature;

import com.mojang.serialization.Codec;
import com.mojang.serialization.codecs.RecordCodecBuilder;
import net.minecraft.block.BlockState;
import net.minecraft.world.gen.feature.FeatureConfig;

public final class ShrubFeatureConfig implements FeatureConfig {
    public static final Codec<ShrubFeatureConfig> CODEC = RecordCodecBuilder.create(
        instance -> instance.group(
            BlockState.CODEC.fieldOf("log").forGetter(ShrubFeatureConfig::getLog),
            BlockState.CODEC.fieldOf("leaves").forGetter(ShrubFeatureConfig::getLeaves),
            Codec.INT.fieldOf("base_height").forGetter(ShrubFeatureConfig::getBaseHeight),
            Codec.INT.fieldOf("extra_height").forGetter(ShrubFeatureConfig::getExtraHeight),
            Codec.FLOAT.fieldOf("extra_height_chance").forGetter(ShrubFeatureConfig::getExtraHeightChance)
        ).apply(instance, ShrubFeatureConfig::new)
    );

    final BlockState log;
    final BlockState leaves;
    final int baseHeight;
    final int extraHeight;
    final float extraHeightChance;

    public ShrubFeatureConfig(BlockState log, BlockState leaves, int baseHeight, int extraHeight, float extraHeightChance) {
        this.log = log;
        this.leaves = leaves;
        this.baseHeight = baseHeight;
        this.extraHeight = extraHeight;
        this.extraHeightChance = extraHeightChance;
    }

    public BlockState getLog() {
        return log;
    }

    public BlockState getLeaves() {
        return leaves;
    }

    public int getBaseHeight() {
        return baseHeight;
    }

    public int getExtraHeight() {
        return extraHeight;
    }

    public float getExtraHeightChance() {
        return extraHeightChance;
    }
}

File: src/main/kotlin/juuxel/woodsandmires/feature/ShrubFeatureConfig.kt
package juuxel.woodsandmires.feature

import com.mojang.serialization.Codec
import com.mojang.serialization.codecs.RecordCodecBuilder
import net.minecraft.block.BlockState
import net.minecraft.world.gen.feature.FeatureConfig

data class ShrubFeatureConfig(
    val log: BlockState,
    val leaves: BlockState,
    val baseHeight: Int,
    val extraHeight: Int,
    val extraHeightChance: Float
) : FeatureConfig {
    companion object {
        val CODEC: Codec<ShrubFeatureConfig> = RecordCodecBuilder.create { instance ->
            instance.group(
                BlockState.CODEC.fieldOf("log").forGetter { it.log },
                BlockState.CODEC.fieldOf("leaves").forGetter { it.leaves },
                Codec.INT.fieldOf("base_height").forGetter { it.baseHeight },
                Codec.INT.fieldOf("extra_height").forGetter { it.extraHeight },
                Codec.FLOAT.fieldOf("extra_height_chance").forGetter { it.extraHeightChance }
            ).apply(instance, ::ShrubFeatureConfig)
        }
    }
}
